### PR TITLE
Nameless fields

### LINKED ISSUE #4738: Missing Declaration in Struct causes translate-c to Generate Invalid Zig
Some headers (Windows) use structs with missing declarations. When importing the Windows headers, you can `#define NONAMELESSUNION` to replace all the nameless unions/structs with dummy values. Discussion in  #985, #1214, and #4081 might be relevant to how to implement this.

There are four different cases I tried in Zig:

* No names defined (valid Zig code, but no elements)
* One name defined (valid Zig code with the single named element)
* Unnamed struct as a member (invalid Zig code)
* Nested struct declaration without a name generates (valid Zig code, generated name)

Based on the fourth case, I believe that the previous three should all be generating unnamed elements instead of excluding them to be consistent. One of the referenced issues mentions that Zig can re-order the struct contents, so it may not be that simple to maintain compatibility. 

No names defined generates valid Zig code with no elements:
```c
typedef struct NONAMES
{
    int;
    long;
} NONAMES;
```

```zig
pub const struct_NONAMES = extern struct {};
pub const NONAMES = struct_NONAMES;
```

One name generates valid Zig code with the single named element:
```c
typedef struct ONENAME
{
    int;
    long b;
} ONENAME;
```

```zig
pub const struct_ONENAME = extern struct {
    b: c_long,
};
pub const ONENAME = struct_ONENAME;
```

With an unnamed struct, invalid Zig code is generated.
```c
typedef struct NAMED
{
    long name;
} NAMED;

typedef struct ONENAMEWITHSTRUCT
{
    NAMED;
    long b;
} ONENAMEWITHSTRUCT;
```

```zig
pub const struct_NAMED = extern struct {
    name: c_long,
};
pub const NAMED = struct_NAMED;
pub const struct_ONENAMEWITHSTRUCT = extern struct {
    : struct_NAMED,
    b: c_long,
};
```

A nested struct declaration without a name generates valid Zig code:
```c
typedef struct ONENAMEWITHSTRUCT
{
    struct {
        long name;
    };
    long b;
} ONENAMEWITHSTRUCT;
```

```zig
const struct_unnamed_2 = extern struct {
    name: c_long,
};
pub const struct_ONENAMEWITHSTRUCT = extern struct {
    unnamed_1: struct_unnamed_2,
    b: c_long,
};
```

### PR DESCRIPTION
**Pull request by SuperAuguste**
Closes #4738.

I'm not sure if this is the right approach or not. This makes nameless struct fields consistent by converting
```c
typedef struct NAMED
{
    long name;
} NAMED;

typedef struct ONENAMEWITHSTRUCT
{
    NAMED;
    long b;
} ONENAMEWITHSTRUCT;
```

to

```zig
pub const struct_NAMED = extern struct {
    name: c_long,
};
pub const NAMED = struct_NAMED;
pub const struct_ONENAMEWITHSTRUCT = extern struct {
    b: c_long,
};
```

instead of

```zig
pub const struct_NAMED = extern struct {
    name: c_long,
};
pub const NAMED = struct_NAMED;
pub const struct_ONENAMEWITHSTRUCT = extern struct {
    : struct_NAMED,
    b: c_long,
};
```

### CODE REVIEW

**Comment by SuperAuguste** [comment]
The issue's author stated that it'd be better to use dummy names instead of entirely removing nameless fields, so if that's the common consensus, I can rework this pull request to that end.

File: src-self-hosted/translate_c.zig [hunk 1]
```diff
@@ -815,6 +815,9 @@ fn transRecordDecl(c: *Context, record_decl: *const ZigClangRecordDecl) Error!?*
                 raw_name = try std.fmt.allocPrint(c.a(), "unnamed_{}", .{c.getMangle()});
                 is_anon = true;
             }
+            if (raw_name.len == 0) {
```
Comment (Vexu) [inline src-self-hosted/translate_c.zig:818]: On msvc
```c
typedef struct NAMED
{
    long name;
} NAMED;

typedef struct ONENAMEWITHSTRUCT
{
    NAMED;
    long b;
} ONENAMEWITHSTRUCT;
```
should generate
```zig
pub const struct_NAMED = extern struct {
    name: c_long,
};
pub const NAMED = struct_NAMED;
pub const struct_ONENAMEWITHSTRUCT = extern struct {
    unnamed_1: struct_NAMED,
    b: c_long,
};
```
which should happen if you just move `or raw_name.len == 0` to line 814.

You can add a test for this by using:
```zig
if (std.Target.current.abi == .msvc) {
    ...
} else {
    ...
}

Comment (SuperAuguste) [reply src-self-hosted/translate_c.zig:818]: Sounds good!

**Comment by andrewrk** [comment]
Sorry about the master branch breakage - fixed in ed23dad4877cd93fd684852b17e0e554ef71e297.

**Pull request merged**